rust/lance-io/src/object_store/gcs_wrapper.rs

```diff
@@ -0,0 +1,458 @@
+                    Poll::Ready(Ok(multipart_id)) => {
+                        let futures = FuturesUnordered::new();
+                        let multipart_id = Arc::new(multipart_id);
+
+                        futures.push(Self::put_next_part(
+                            &mut_self.path,
+                            &mut_self.store,
+                            &mut mut_self.buffer,
+                            &mut 0,
+                            &multipart_id,
+                        ));
+
+                        mut_self.state = UploadState::InProgress {
+                            multipart_id,
+                            part_idx: 1, // We just used 0
+                            futures,
+                            part_ids: Vec::new(),
+                        };
+                    }
+                    Poll::Ready(Err(e)) => {
+                        return Err(std::io::Error::new(std::io::ErrorKind::Other, e))
+                    }
+                    Poll::Pending => break,
+                },
+                UploadState::InProgress {
+                    futures,
+                    part_ids,
+                    multipart_id,
+                    ..
+                } => {
+                    while let Poll::Ready(Some(res)) = futures.poll_next_unpin(cx) {
+                        match res {
+                            Ok((part_idx, part_id)) => {
+                                let total_parts = part_ids.len();
+                                part_ids.resize(total_parts.max(part_idx as usize + 1), None);
+                                part_ids[part_idx as usize] = Some(part_id);
+                            }
+                            Err(UploadPutError {
+                                source: OSError::Generic { source, .. },
+                                part_idx,
+                                buffer,
+                            }) if source.to_string().contains("Connection reset by peer")
+                                && mut_self.connection_resets < 20 =>
+                            {
+                                // Retry, but only up to 20 of them.
+                                mut_self.connection_resets += 1;
+
+                                // Resubmit with random jitter
+                                let sleep_time_ms = rand::thread_rng().gen_range(2_000..8_000);
+                                let sleep_time = std::time::Duration::from_millis(sleep_time_ms);
+
+                                futures.push(Self::put_part(
+                                    mut_self.path.clone(),
+                                    mut_self.store.clone(),
+                                    buffer,
+                                    part_idx,
+                                    multipart_id.clone(),
+                                    Some(sleep_time),
+                                ));
+                            }
```

Comment:
This is an experimental retry mechanism. I noticed I would sometimes hit these errors when uploading. We'll see whether these continue to help in production use cases and if we need to do anything to make it more robust or configurable.

Reply:
Is there any retry mechanism in the inner store?

This is very nice btw.